packages/loader/container-loader/src/pendingLocalStateStore.ts

```diff
@@ -0,0 +1,160 @@
+ * Licensed under the MIT License.
+ */
+
+import type { ISequencedDocumentMessage } from "@fluidframework/driver-definitions/internal";
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+
+import type {
+	IPendingContainerState,
+	SerializedSnapshotInfo,
+} from "./serializedStateManager.js";
+import { getAttachedContainerStateFromSerializedContainer } from "./utils.js";
+
+/**
+ * A Map-like store for managing pending local container states from attached containers.
+ * Optimizes storage by deduplicating shared resources across stored states.
+ *
+ * @example
+ * ```typescript
+ * const store = new PendingLocalStateStore<string>();
+ *
+ * // Store pending state
+ * const pendingState = await attachedContainer.getPendingLocalState();
+ * store.set("session1", pendingState);
+ *
+ * // Load from stored state
+ * const restored = store.get("session1");
+ * const newContainer = await loadFrozenContainerFromPendingState({
+ *   pendingLocalState: restored,
+ *   // ... other loader options
+ * });
+ * ```
+ *
+ * @remarks
+ * Only use with attached containers from the same URL. Only store strings
+ * returned by `container.getPendingLocalState()`.
+ *
+ * @typeParam TKey - The type of keys used to identify stored states
+ *
+ * @legacy @alpha
+ */
+export class PendingLocalStateStore<TKey> {
+	#firstUrl: string | undefined;
+	readonly #pendingStates = new Map<TKey, IPendingContainerState>();
+	readonly #savedOps: Record<number, ISequencedDocumentMessage> = {};
+	readonly #blobs: Record<string, string> = {};
+	readonly #loadingGroups: Record<string, SerializedSnapshotInfo> = {};
+
+	/**
+	 * Removes all stored pending states.
+	 */
+	clear(): void {
+		return this.#pendingStates.clear();
+	}
+
+	/**
+	 * Removes the pending state for the specified key.
+	 *
+	 * @param key - The key to remove
+	 * @returns `true` if the state existed and was removed, `false` otherwise
+	 */
+	delete(key: TKey): boolean {
+		return this.#pendingStates.delete(key);
+	}
+
+	/**
+	 * Retrieves the serialized pending state for the specified key.
+	 *
+	 * @param key - The key to retrieve
+	 * @returns The serialized state as a JSON string, or `undefined` if not found
+	 */
+	get(key: TKey): string | undefined {
+		return JSON.stringify(this.#pendingStates.get(key));
+	}
+
+	/**
+	 * Checks whether a pending state exists for the specified key.
+	 */
+	has(key: TKey): boolean {
+		return this.#pendingStates.has(key);
+	}
+
+	/**
+	 * Stores a pending state from `container.getPendingLocalState()`.
+	 *
+	 * @param key - The key to associate with the state
+	 * @param pendingLocalState - String returned by `getPendingLocalState()` from an attached container
+	 * @returns This store instance for method chaining
+	 *
+	 * @throws When storing states from different container URLs
+	 */
+	set(key: TKey, pendingLocalState: string): this {
+		const state = getAttachedContainerStateFromSerializedContainer(pendingLocalState);
```

Comment:
maybe we could add a "isPendingContainerState" method here to strengthen  validation. Current method just parses the string blindly which I don't think it's enough.

Reply:
does a method like that exist? i agree it is good idea, and i have some ideas about making validation better, but didn't want to mix them into this pr